apps/frontend/app/(website)/studio/main/ASTViewer/AboristViewer.tsx

```diff
@@ -27,12 +29,85 @@ interface Props {
   type: EditorType;
 }
 
-const transformTreeData = (node: TreeNode): TreeNode => ({
-  ...node,
-  children: node.children?.length
-    ? node.children.map(transformTreeData)
-    : undefined,
-});
+const removeEmptyChildren = (node: TreeNode): TreeNode => {
+  const transformedChildren =
+    node.children
+      ?.map(removeEmptyChildren)
+      .filter((child) => child.children?.length || !child.children) || [];
+
+  return {
+    ...node,
+    children: transformedChildren.length > 0 ? transformedChildren : undefined,
+  };
+};
+
+const parseNodeData =
+  (node: TreeNode, identifier: string) => (param, index) => ({
+    id: `${node.id}-${identifier}-${index}`,
+    actualNode: param,
+    label: param.type,
+    start: param.start,
+    end: param.end,
+  });
+const getPosition = (node: TreeNode) => `${node.start}-${node.end}`;
+const transformTreeData = (node: TreeNode, relation?: string): TreeNode => {
+  const result: TreeNode = {
+    ...node,
+    relation,
+    children: [],
+  };
+
+  if (node.actualNode.type === "CallExpression") {
+    const args = node.actualNode.arguments.map(parseNodeData(node, "arg"));
+    const argumentsIds = args.map(getPosition);
+    result.children =
+      node.children.map((child) => {
+        const isParam = argumentsIds.includes(getPosition(child));
+        return transformTreeData(child, isParam ? "argument" : child.label);
+      }) || [];
+  } else if (node.actualNode.type === "FunctionDeclaration") {
```

Comment:
prefer early returns over nested conditionals if it's possible to use them please

Reply:
@r4zendev what do you mean by nested if? We have just one level of nesting here (it is else if, not if within an if)

Reply:
yeah, i phrased it incorrectly. in general, early returns are easier to read than multiple `else if` statements. here, on line 99 you return result that is being mutated in each conditional. the alternative is to return result on lines 68, 85, 97 and replace `else if` with `if`, remove `else`. i could argue that this makes the code more readable. if you agree, let's replace it with early returns. i agree with both, but prefer my option slightly more.

Reply:
Updates added. When it comes to `comments` I made it mutable as we are traversing the tree so it simplifies the code a lot this way